packages/muelu/src/Misc/MueLu_InterfaceAggregationFactory_def.hpp

```diff
@@ -0,0 +1,196 @@
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact
+//                    Jonathan Hu       (jhu@sandia.gov)
+//                    Andrey Prokopenko (aprokop@sandia.gov)
+//                    Ray Tuminaro      (rstumin@sandia.gov)
+//
+// ***********************************************************************
+//
+// @HEADER
+#ifndef MUELU_INTERFACEAGGREGATIONFACTORY_DEF_HPP_
+#define MUELU_INTERFACEAGGREGATIONFACTORY_DEF_HPP_
+
+#include "MueLu_InterfaceAggregationFactory_decl.hpp"
+
+#include "MueLu_Level.hpp"
+#include "MueLu_Aggregates.hpp"
+#include "MueLu_Monitor.hpp"
+#include "Xpetra_MapFactory.hpp"
+
+namespace MueLu
+{
+
+template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+InterfaceAggregationFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::InterfaceAggregationFactory()
+{
+}
+
+template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+RCP<const ParameterList> InterfaceAggregationFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::GetValidParameterList() const
+{
+  RCP<ParameterList> validParamList = rcp(new ParameterList());
+
+  validParamList->set<RCP<const FactoryBase>>("A", null, "Generating factory of A");
+  validParamList->set<RCP<const FactoryBase>>("Aggregates", null, "Generating factory of the Aggregates (for block 0,0)");
+  validParamList->set<RCP<const FactoryBase>>("DualNodeID2PrimalNodeID", null, "Generating factory of the DualNodeID2PrimalNodeID map");
+
+  validParamList->set<LocalOrdinal>("number of DOFs per dual node", Teuchos::ScalarTraits<LocalOrdinal>::one(), "Number of DOFs per dual node");
+
+  /*
+    DualNodeID2PrimalNodeID represents the mapping between the Dual Node IDs to the Primal Node IDs.
+    This map makes one assumption: the number of dof per dual node is a constant and has to be set in 
+    the parameter list as "number of DOFs per dual node".
+  */
+  validParamList->set<RCP<std::map<LocalOrdinal, LocalOrdinal>>>("DualNodeID2PrimalNodeID - level 0", null, "");
+  return validParamList;
+}
+
+template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+void InterfaceAggregationFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::DeclareInput(Level &currentLevel) const
+{
+  Input(currentLevel, "A");
+  Input(currentLevel, "Aggregates");
+
+  if (currentLevel.GetLevelID() != 0)
+  {
+    Input(currentLevel, "DualNodeID2PrimalNodeID");
+  }
+}
+
+template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+void InterfaceAggregationFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::Build(Level &currentLevel) const
+{
+  typedef Xpetra::Map<LocalOrdinal, GlobalOrdinal, Node> Map;
+  typedef Aggregates<LocalOrdinal, GlobalOrdinal, Node> Aggregates;
+  typedef std::map<LocalOrdinal, LocalOrdinal> Dual2Primal_type;
+  typedef Xpetra::MapFactory<LocalOrdinal, GlobalOrdinal, Node> MapFactory;
```

Comment:
consider switching `typedef` to `using`, it makes @mhoemmen happy ; )

Reply:
Also I think these are already defined in the `UseShortNamesOrdinal` header, I think you should check and try to remove these as they might be redefinitions...

Reply:
Finally if you get tired of typing `LocalOrdinal` you could just switch to `LO`...

Reply:
Concerning `UseShortNamesOrdinal`, it was what I thought too but it is not working if I remove one of those `typedef`. I will, however, use `using` instead of `typedef`! Thanks!